File: scripts/go_gen_p4_const.py
# Copyright 2022-present Open Networking Foundation
#
# SPDX-License-Identifier: Apache-2.0

import argparse
import google.protobuf.text_format as tf
import re
from p4.config.v1 import p4info_pb2

copyright = '''/*
 * Copyright 2022-present Open Networking Foundation
 *
 * SPDX-License-Identifier: Apache-2.0
 */
'''

PKG_FMT = 'package %s'

CONST_OPEN = '''//noinspection GoSnakeCaseUsage
    const ('''
CONST_CLOSE = ')'


UINT32 = 'uint32'
EMPTY_STR = ''

HF_VAR_PREFIX = 'Hdr_'
TBL_VAR_PREFIX = 'Table_'
CTR_VAR_PREFIX = 'Counter_'
DIRCTR_VAR_PREFIX = 'DirectCounter_'
ACT_VAR_PREFIX = "Action_"
ACTPARAM_VAR_PREFIX = "ActionParam_"
ACTPROF_VAR_PREFIX = "ActionProfile_"
PACKETMETA_VAR_PREFIX = "PacketMeta_"
MTR_VAR_PREFIX = "Meter_"

class ConstantClassGenerator(object):
    header_fields = dict()
    tables = set()
    counters = set()
    direct_counters = set()
    actions = set()
    action_params = dict()
    action_profiles = set()
    packet_metadata = set()
    meters = set()

    # https://stackoverflow.com/questions/1175208/elegant-python-function-to-convert-camelcase-to-snake-case
    def convert_camel_to_all_caps(self, name):
        s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
        s1 = re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).title()
        return s1.replace('_', '').replace('.', '')

    def __init__(self, pkg_name):
        self.package_name = PKG_FMT % (pkg_name, )

    def parse(self, p4info):
        for tbl in p4info.tables:
            for mf in tbl.match_fields:
                try:
                    self.header_fields[tbl.preamble.name]
                except KeyError:
                    self.header_fields[tbl.preamble.name] = set()
                finally:
                    self.header_fields[tbl.preamble.name].add((mf.name, mf.id))

            self.tables.add((tbl.preamble.name, tbl.preamble.id))

        for ctr in p4info.counters:
            self.counters.add((ctr.preamble.name, ctr.preamble.id))

        for dir_ctr in p4info.direct_counters:
            self.direct_counters.add((dir_ctr.preamble.name, dir_ctr.preamble.id))

        for act in p4info.actions:
            self.actions.add((act.preamble.name, act.preamble.id))

            for param in act.params:
                try:
                    self.action_params[act.preamble.name]
                except KeyError as e:
                    self.action_params[act.preamble.name] = set()
                finally:
                    self.action_params[act.preamble.name].add((param.name, param.id))

        for act_prof in p4info.action_profiles:
            self.action_profiles.add((act_prof.preamble.name, act_prof.preamble.id))

        for cpm in p4info.controller_packet_metadata:
            for mta in cpm.metadata:
                self.packet_metadata.add((mta.name, mta.id))
        for mtr in p4info.meters:
            self.meters.add((mtr.preamble.name, mtr.preamble.id))

    def const_line(self, obj_type, name, id, var_type=UINT32, table_prefix=None):
        var_name = self.convert_camel_to_all_caps(name)
        if table_prefix:
            var_name = self.convert_camel_to_all_caps(table_prefix) + "_" + var_name
        var_name = obj_type+var_name
        base_line = "\t%s %s = %i"
        line = base_line % (var_name, var_type, id)
        return line

    def generate_go(self):
        lines = list()
        lines.append(copyright)
        lines.append(self.package_name)
        lines.append(CONST_OPEN)

        if len(self.header_fields) != 0:
            lines.append('    // Header field IDs')
        for (tbl, hfs) in self.header_fields.items():
            for hf in hfs:
                lines.append(self.const_line(HF_VAR_PREFIX, hf[0], hf[1], table_prefix=tbl))

        if len(self.tables) != 0:
            lines.append('    // Table IDs')
        for tbl in self.tables:
            lines.append(self.const_line(TBL_VAR_PREFIX, tbl[0], tbl[1]))

        if len(self.counters) != 0:
            lines.append('    // Indirect Counter IDs')
        for ctr in self.counters:
            lines.append(self.const_line(CTR_VAR_PREFIX, ctr[0], ctr[1]))

        if len(self.direct_counters) != 0:
            lines.append('    // Direct Counter IDs')
        for dctr in self.direct_counters:
            lines.append(self.const_line(DIRCTR_VAR_PREFIX, dctr[0], dctr[1]))

        if len(self.actions) != 0:
            lines.append('    // Action IDs')
        for act in self.actions:
            lines.append(self.const_line(ACT_VAR_PREFIX, act[0], act[1]))

        if len(self.action_params) != 0:
            lines.append('    // Action Param IDs')
        for (tbl, act_prms) in self.action_params.items():
            for act_prm in act_prms:
                lines.append(self.const_line(ACTPARAM_VAR_PREFIX, act_prm[0], act_prm[1], table_prefix=tbl))

        if len(self.action_profiles) != 0:
            lines.append('    // Action Profile IDs')
        for act_prof in self.action_profiles:
            lines.append(self.const_line(ACTPROF_VAR_PREFIX, act_prof[0], act_prof[1]))

        if len(self.packet_metadata) != 0:
            lines.append('    // Packet Metadata IDs')
        for pmeta in self.packet_metadata:
            if not pmeta[0].startswith("_"):
                lines.append(self.const_line(PACKETMETA_VAR_PREFIX, pmeta[0], pmeta[1]))

        if len(self.meters) != 0:
            lines.append('    // Meter IDs')
        for mtr in self.meters:
            lines.append(self.const_line(MTR_VAR_PREFIX, mtr[0], mtr[1]))
        lines.append(CONST_CLOSE)
        # end of class

        return '\n'.join(lines)

def main():
    parser = argparse.ArgumentParser(prog='go-gen-p4-const',
                                     description='P4Info to Go constant generator.')
    parser.add_argument('-o', '--output', help='path to output file', default='-')
    parser.add_argument('-p', '--p4info', help='path to p4info file (text format)')
    args = parser.parse_args()

    p4info_file = args.p4info
    output_file = args.output

    pieces = args.output.split('/')
    pkg_name = pieces[-2] if len(pieces) > 1 else 'undefined'

    p4info = p4info_pb2.P4Info()
    with open(p4info_file, 'r') as input_file:
        s = input_file.read()
        tf.Merge(s, p4info)

    gen = ConstantClassGenerator(pkg_name)
    gen.parse(p4info)
    go_code = gen.generate_go()

    if output_file == '-':
        # std output
        print(go_code)
    else:
        with open(output_file, 'w') as output_file:
            output_file.write(go_code)


if __name__ == '__main__':
    main()

File: scripts/go_gen_p4_const.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2022-present Open Networking Foundation

package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"strings"

	"github.com/ettle/strcase"
	"github.com/golang/protobuf/proto"
	p4ConfigV1 "github.com/p4lang/p4runtime/go/p4/config/v1"
)

const (
	p4infoPath = "conf/p4/bin/p4info.txt"

	defaultPackageName = "p4constants"
	// copyrightHeader uses raw strings to avoid issues with reuse
	copyrightHeader = `// SPDX-License-Identifier: Apache-2.0
// Copyright 2022-present Open Networking Foundation
`

	constOpen        = "const (\n"
	mapFormatString  = "%v:\"%v\",\n"
	listFormatString = "%v,\n"
	constOrVarClose  = ")\n"

	idTypeString   = "uint32"
	sizeTypeString = "uint64"

	hfVarPrefix         = "Hdr_"
	tblVarPrefix        = "Table_"
	ctrVarPrefix        = "Counter_"
	ctrSizeVarPrefix    = "CounterSize_"
	dirCtrVarPrefix     = "DirectCounter_"
	actVarPrefix        = "Action_"
	actparamVarPrefix   = "ActionParam_"
	actprofVarPrefix    = "ActionProfile_"
	packetmetaVarPrefix = "PacketMeta_"
	mtrVarPrefix        = "Meter_"
	mtrSizeVarPrefix    = "MeterSize_"

	tableMapFunc          = "func GetTableIDToNameMap() map[uint32]string {\n return map[uint32]string {\n"
	tableListFunc         = "func GetTableIDList() []uint32 {\n return []uint32 {\n"
	actionMapFunc         = "func GetActionIDToNameMap() map[uint32]string {\n return map[uint32]string {\n"
	actionListFunc        = "func GetActionIDList() []uint32 {\n return []uint32 {\n"
	counterMapFunc        = "func GetCounterIDToNameMap() map[uint32]string {\n return map[uint32]string {\n"
	counterListFunc       = "func GetCounterIDList() []uint32 {\n return []uint32 {\n"
	directCounterMapFunc  = "func GetDirectCounterIDToNameMap() map[uint32]string {\n return map[uint32]string {\n"
	directCounterListFunc = "func GetDirectCounterIDList() []uint32 {\n return []uint32 {\n"
	actionProfileMapFunc  = "func GetActionProfileIDToNameMap() map[uint32]string {\n return map[uint32]string {\n"
	actionProfileListFunc = "func GetActionProfileIDList() []uint32 {\n return []uint32 {\n"
	pktMetadataMapFunc    = "func GetPacketMetadataIDToNameMap() map[uint32]string {\n return map[uint32]string {\n"
	pktMetadataListFunc   = "func GetPacketMetadataIDList() []uint32 {\n return []uint32 {\n"
	metersMapFunc         = "func GetMeterIDToNameMap() map[uint32]string {\n return map[uint32]string {\n"
	metersListFunc        = "func GetMeterIDList() []uint32 {\n return []uint32 {\n"
)

func emitEntityConstant(prefix string, p4EntityName string, id uint32) string {
	// see: https://go.dev/ref/spec#Identifiers
	p4EntityName = prefix + "_" + p4EntityName
	p4EntityName = strings.Replace(p4EntityName, ".", "_", -1)
	p4EntityName = strcase.ToPascal(p4EntityName)
	return fmt.Sprintf("%s \t %s = %v\n", p4EntityName, idTypeString, id)
}

func emitEntitySizeConstant(prefix string, p4EntityName string, id int64) string {
	// see: https://go.dev/ref/spec#Identifiers
	p4EntityName = prefix + "_" + p4EntityName
	p4EntityName = strings.Replace(p4EntityName, ".", "_", -1)
	p4EntityName = strcase.ToPascal(p4EntityName)
	return fmt.Sprintf("%s \t %s = %v\n", p4EntityName, sizeTypeString, id)
}

func generateTables(p4info *p4ConfigV1.P4Info) string {
	mapBuilder, listBuilder := strings.Builder{}, strings.Builder{}

	mapBuilder.WriteString(tableMapFunc)
	listBuilder.WriteString(tableListFunc)
	for _, element := range p4info.GetTables() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		mapBuilder.WriteString(fmt.Sprintf(mapFormatString, ID, name))
		listBuilder.WriteString(fmt.Sprintf(listFormatString, ID))
	}
	mapBuilder.WriteString("}\n}\n\n")
	listBuilder.WriteString("}\n}\n\n") //Close func declaration

	return mapBuilder.String() + listBuilder.String()
}

func generateActions(p4info *p4ConfigV1.P4Info) string {
	mapBuilder, listBuilder := strings.Builder{}, strings.Builder{}

	mapBuilder.WriteString(actionMapFunc)
	listBuilder.WriteString(actionListFunc)
	for _, element := range p4info.GetActions() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		mapBuilder.WriteString(fmt.Sprintf(mapFormatString, ID, name))
		listBuilder.WriteString(fmt.Sprintf(listFormatString, ID))
	}
	mapBuilder.WriteString("}\n}\n\n")
	listBuilder.WriteString("}\n}\n\n") //Close func declarations

	return mapBuilder.String() + listBuilder.String()
}

func generateIndirectCounters(p4info *p4ConfigV1.P4Info) string {
	mapBuilder, listBuilder := strings.Builder{}, strings.Builder{}

	mapBuilder.WriteString(counterMapFunc)
	listBuilder.WriteString(counterListFunc)
	for _, element := range p4info.GetCounters() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		mapBuilder.WriteString(fmt.Sprintf(mapFormatString, ID, name))
		listBuilder.WriteString(fmt.Sprintf(listFormatString, ID))
	}
	mapBuilder.WriteString("}\n}\n\n")
	listBuilder.WriteString("}\n}\n\n") //Close func declarations

	return mapBuilder.String() + listBuilder.String()
}

func generateDirectCounters(p4info *p4ConfigV1.P4Info) string {
	mapBuilder, listBuilder := strings.Builder{}, strings.Builder{}

	mapBuilder.WriteString(directCounterMapFunc)
	listBuilder.WriteString(directCounterListFunc)
	for _, element := range p4info.GetDirectCounters() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		mapBuilder.WriteString(fmt.Sprintf(mapFormatString, ID, name))
		listBuilder.WriteString(fmt.Sprintf(listFormatString, ID))
	}
	mapBuilder.WriteString("}\n}\n\n")
	listBuilder.WriteString("}\n}\n\n") //Close declarations

	return mapBuilder.String() + listBuilder.String()
}

func generateMeters(p4info *p4ConfigV1.P4Info) string {
	mapBuilder, listBuilder := strings.Builder{}, strings.Builder{}

	// Meters
	mapBuilder.WriteString(metersMapFunc)
	listBuilder.WriteString(metersListFunc)
	for _, element := range p4info.GetMeters() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		mapBuilder.WriteString(fmt.Sprintf(mapFormatString, ID, name))
		listBuilder.WriteString(fmt.Sprintf(listFormatString, ID))
	}
	mapBuilder.WriteString("}\n}\n\n")
	listBuilder.WriteString("}\n}\n\n") //Close declarations

	return mapBuilder.String() + listBuilder.String()
}

func generateActionProfiles(p4info *p4ConfigV1.P4Info) string {
	mapBuilder, listBuilder := strings.Builder{}, strings.Builder{}

	mapBuilder.WriteString(actionProfileMapFunc)
	listBuilder.WriteString(actionProfileListFunc)
	for _, element := range p4info.GetActionProfiles() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		mapBuilder.WriteString(fmt.Sprintf(mapFormatString, ID, name))
		listBuilder.WriteString(fmt.Sprintf(listFormatString, ID))
	}
	mapBuilder.WriteString("}\n}\n\n")
	listBuilder.WriteString("}\n}\n\n") //Close declarations

	return mapBuilder.String() + listBuilder.String()
}

func generatePacketMetadata(p4info *p4ConfigV1.P4Info) string {
	mapBuilder, listBuilder := strings.Builder{}, strings.Builder{}

	mapBuilder.WriteString(pktMetadataMapFunc)
	listBuilder.WriteString(pktMetadataListFunc)
	for _, element := range p4info.GetControllerPacketMetadata() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		mapBuilder.WriteString(fmt.Sprintf(mapFormatString, ID, name))
		listBuilder.WriteString(fmt.Sprintf(listFormatString, ID))
	}
	mapBuilder.WriteString("}\n}\n\n")
	listBuilder.WriteString("}\n}\n\n") //Close declarations

	return mapBuilder.String() + listBuilder.String()
}

func generateConstants(p4info *p4ConfigV1.P4Info) string {
	constBuilder := strings.Builder{}

	constBuilder.WriteString(constOpen)

	//HeaderField IDs
	constBuilder.WriteString("// HeaderFields\n")
	for _, element := range p4info.GetTables() {
		for _, matchField := range element.MatchFields {
			tableName, name := element.GetPreamble().GetName(), matchField.GetName()

			constBuilder.WriteString(emitEntityConstant(hfVarPrefix+tableName, name, matchField.GetId()))
		}
	}
	// Tables
	constBuilder.WriteString("// Tables\n")
	for _, element := range p4info.GetTables() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		constBuilder.WriteString(emitEntityConstant(tblVarPrefix, name, ID))
	}

	// Actions
	constBuilder.WriteString("// Actions\n")
	for _, element := range p4info.GetActions() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		constBuilder.WriteString(emitEntityConstant(actVarPrefix, name, ID))
	}

	// Action Param IDs
	constBuilder.WriteString("// ActionParams\n")
	for _, element := range p4info.GetActions() {
		for _, actionParam := range element.GetParams() {
			actionName, name := element.GetPreamble().GetName(), actionParam.GetName()

			constBuilder.WriteString(emitEntityConstant(actparamVarPrefix+actionName, name, actionParam.GetId()))
		}
	}

	// Indirect Counters
	constBuilder.WriteString("// IndirectCounters\n")
	for _, element := range p4info.GetCounters() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		constBuilder.WriteString(emitEntityConstant(ctrVarPrefix, name, ID))
		constBuilder.WriteString(emitEntitySizeConstant(ctrSizeVarPrefix, name, element.GetSize()))
	}

	// Direct Counters
	constBuilder.WriteString("// DirectCounters\n")
	for _, element := range p4info.GetDirectCounters() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		constBuilder.WriteString(emitEntityConstant(dirCtrVarPrefix, name, ID))
	}

	// Action profiles
	constBuilder.WriteString("// ActionProfiles\n")
	for _, element := range p4info.GetActionProfiles() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		constBuilder.WriteString(emitEntityConstant(actprofVarPrefix, name, ID))
	}

	// Packet metadata
	constBuilder.WriteString("// PacketMetadata\n")
	for _, element := range p4info.GetControllerPacketMetadata() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		constBuilder.WriteString(emitEntityConstant(packetmetaVarPrefix, name, ID))
	}

	// Meters
	constBuilder.WriteString("// Meters\n")
	for _, element := range p4info.GetMeters() {
		name, ID := element.GetPreamble().GetName(), element.GetPreamble().GetId()

		constBuilder.WriteString(emitEntityConstant(mtrVarPrefix, name, ID))
		constBuilder.WriteString(emitEntitySizeConstant(mtrSizeVarPrefix, name, element.GetSize()))
	}

	constBuilder.WriteString(constOrVarClose + "\n")

	return constBuilder.String()
}

func getP4Config(p4infopath string) *p4ConfigV1.P4Info {
	p4infoBytes, err := ioutil.ReadFile(p4infopath)
	if err != nil {
		panic(fmt.Sprintf("Could not read P4Info file: %v", err))
	}

	var p4info p4ConfigV1.P4Info

	err = proto.UnmarshalText(string(p4infoBytes), &p4info)
	if err != nil {
		panic("Could not parse P4Info file")
	}

	return &p4info
}

func main() {
	p4infoPath := flag.String("p4info", p4infoPath, "Path of the p4info file")
	outputPath := flag.String("output", "-", "Default will print to Stdout")
	packageName := flag.String("package", defaultPackageName, "Set the package name")

	flag.Parse()

	p4config := getP4Config(*p4infoPath)

	headerBuilder := strings.Builder{}

	headerBuilder.WriteString(copyrightHeader + "\n")
	headerBuilder.WriteString(fmt.Sprintf("package %s\n", *packageName))

	headerBuilder.WriteString(generateConstants(p4config))

	headerBuilder.WriteString(generateTables(p4config))
	headerBuilder.WriteString(generateActions(p4config))
	headerBuilder.WriteString(generateDirectCounters(p4config))
	headerBuilder.WriteString(generateIndirectCounters(p4config))
	headerBuilder.WriteString(generateActionProfiles(p4config))
	headerBuilder.WriteString(generatePacketMetadata(p4config))
	headerBuilder.WriteString(generateMeters(p4config))

	result := headerBuilder.String()

	if *outputPath == "-" {
		fmt.Println(result)
	} else {
		if err := os.WriteFile(*outputPath, []byte(result), 0644); err != nil {
			panic(fmt.Sprintf("Error while creating File: %v", err))
		}
	}
}
